app/src/main/java/com/github/bgabriel998/softwaredevproject/Compass.java

```diff
@@ -60,58 +50,70 @@ public void stop() {
-            updateValues(accMat, event, alpha);
-        }
+            //Matrix to cache the rotation vector
+            float[] rotMatFromVector = new float[16];
 
-        //Get the accelerometer data, use filter to smooth the data
-        if (event.sensor.getType() == Sensor.TYPE_MAGNETIC_FIELD) {
-            updateValues(magMat, event, alpha);
-        }
+            // Convert the rotation-vector to a 4x4 matrix.
+            SensorManager.getRotationMatrixFromVector(rotMatFromVector, orientationVector);
+
+            //Rotates the rotation matrix to be expressed in a different coordinate system
+            SensorManager.remapCoordinateSystem(rotMatFromVector, SensorManager.AXIS_X,
+                    SensorManager.AXIS_Z, rotMat);
+
+            //Compute the device orientation with the rotation matrix
+            SensorManager.getOrientation(rotMat, orientationMat);
 
-        //See https://developer.android.com/reference/android/hardware/SensorManager#getRotationMatrix(float[],%20float[],%20float[],%20float[])
-        boolean success = SensorManager.getRotationMatrix(incMat, rotMat, accMat, magMat);
-        if (success) {
-            float[] orientation = new float[3];
-            SensorManager.getOrientation(incMat, orientation);
-            float heading = (float) Math.toDegrees(orientation[0]);
-            //+90 to use the compass in landscape mode correctly
-            heading = (heading + 360 + 90) % 360;
-
-            float[] orientationVertical = new float[3];
-            SensorManager.getOrientation(rotMat, orientationVertical);
-            float headingV = (float) Math.toDegrees(orientation[2]);
-            //HeadingV to use in landscape mode correctly
-            headingV = (headingV * (-1)) % 360;
+            //Convert values to degrees
+            convertArrToDegrees(orientationMat);
+
+            //Add 360° to only get positive values
+            float headingHorizontal = (orientationMat[0] + 360) % 360;
+
+            //Multiply by -1 to get increasing values when inclining the device
+            //Add 90° to get values from 0° to 180°
+            float headingVertical = (orientationMat[1] * (-1) + 90);
 
             //Update the horizontal and vertical heading
-            if (listener != null) {
-                listener.onNewHeading(heading, headingV);
+            if (compassListener != null) {
+                compassListener.onNewHeading(headingHorizontal, headingVertical);
             }
         }
     }
 
     /**
-     * Update the values of the Matrices
-     * @param mat Output matrice
+     * Convert all values of array to degrees
+     * @param orientationMat Array of values that gets converted from radians to degrees
+     */
+    private void convertArrToDegrees(float[] orientationMat) {
+        for(int i=0; i<orientationMat.length; i++){
+            orientationMat[i] = (float)Math.toDegrees(orientationMat[i]);
+        }
+    }
```

Comment:
Change for functional programming statement
```suggestion
    private void convertArrToDegrees(Float[] orientationMat) {
        Arrays.stream(orientationMat).map(val -> (float) Math.toDegrees(val)).toArray();
    }
```
Don't forget to check the validity of the method. Some imports are also needed

Reply:
Arrays.stream() does not support float[]. I would need to first change the float[] to double[] to then use .map or .foreach which isn't worth it in my opinion.